lib/models/room.js

```diff
@@ -214,43 +292,195 @@ Room.prototype.getAvatarUrl = function(baseUrl, width, height, resizeMethod,
+    if (!this._timelineSupport) {
+        throw Error("timeline support is disabled. Set the 'timelineSupport'" +
+                    " parameter to true when creating MatrixClient to enable" +
+                    " it.");
+    }
+
+    var timeline = new EventTimeline(this.roomId);
+    this._timelines.push(timeline);
+    return timeline;
+};
+
+
+/**
+ * Add events to a timeline
+ *
+ * <p>Will fire "Room.timeline" for each event added.
+ *
  * @param {MatrixEvent[]} events A list of events to add.
- * @param {boolean} toStartOfTimeline True to add these events to the start
+ *
+ * @param {boolean} toStartOfTimeline   True to add these events to the start
  * (oldest) instead of the end (newest) of the timeline. If true, the oldest
  * event will be the <b>last</b> element of 'events'.
+ *
+ * @param {module:models/event-timeline~EventTimeline=} timeline   timeline to
+ *    add events to. If not given, events will be added to the live timeline
+ *
+ * @param {string=} paginationToken   token for the next batch of events
+ *
  * @fires module:client~MatrixClient#event:"Room.timeline"
+ *
  */
-Room.prototype.addEventsToTimeline = function(events, toStartOfTimeline) {
-    var stateContext = toStartOfTimeline ? this.oldState : this.currentState;
+Room.prototype.addEventsToTimeline = function(events, toStartOfTimeline,
+                                              timeline, paginationToken) {
+    if (!timeline) {
+        timeline = this._liveTimeline;
+    }
 
-    function checkForRedaction(redactEvent) {
-        return function(e) {
-            return e.getId() === redactEvent.event.redacts;
-        };
+    if (!toStartOfTimeline && timeline == this._liveTimeline) {
+        // special treatment for live events
+        this._addLiveEvents(events);
+        return;
+    }
+
+    var updateToken = false;
+    for (var i = 0; i < events.length; i++) {
+        var existingTimeline = this._checkExistingTimeline(events[i], timeline,
+                                                           toStartOfTimeline);
+        if (existingTimeline) {
+            // switch to the other timeline
+            timeline = existingTimeline;
+            updateToken = false;
+        } else {
+            this._addEventToTimeline(events[i], timeline, toStartOfTimeline);
+            updateToken = true;
+        }
+    }
+    if (updateToken) {
+        timeline.setPaginationToken(paginationToken, toStartOfTimeline);
+    }
+};
+
+/**
+ * Check if this event is already in a timeline, and join up the timelines if
+ * necessary
+ *
+ * @param {MatrixEvent} event           event to add
+ * @param {EventTimeline} timeline      timeline we think we should add to
+ * @param {boolean} toStartOfTimeline   true if we're adding to the start of
+ *    the timeline
+ * @return {?EventTimeline} the timeline with the event already in, or null if
+ *    none
+ * @private
+ */
+Room.prototype._checkExistingTimeline = function(event, timeline,
```

Comment:
The fact that I can't think of a good name for it is probably a clear indication that it's poorly designed. Maybe it would be better if I just inlined it?

For now I've gone with `_checkForNewEventInExistingTimeline`, but that doesn't reflect the fact that it sometimes has the side-effect of joining the timelines.